### PR TITLE
*: log metamorphic constants at startup rather than initialization

### PR DESCRIPTION
**Pull request by AlexTalks**
While we previously logged all metamorphic constants, which are
initialized to random values under test builds, during module
initialization, this leads to an unfortunate issue when tests are run
via dev/Bazel. Since a Go test binary run by dev/Bazel forks itself after
startup (and crucially, after all modules are initialized) before
starting to run the test, this means test output can show two separate,
sets of metamorphic constants, without *any* way of distinguishing which
constants were from before the process was forked and which are from
after (i.e. the ones used during the test). This makes debugging tests
which flake due to the random value of a metamorphic constant a
nightmare to debug, as the output cannot be relied upon.

This changes the functionality to skip logging metamorphic constants at
initialization, instead modifying them to be logged at test or program startup 
by calling `util.LogMetamorphicVariables()`. This ensures that inapplicable 
metamorphic constants from a pre-forked Go test process are never logged.

Epic: none

Release note: None

### CODE REVIEW

**Comment by AlexTalks** [comment]
Some explanation:

While investigating tests that only failed when certain metamorphic flags were initialized to `true`, I modified the test with `skip.UnderMetamorphic(..)` (in #101882). However, when I would run or stress the test, I would still see the test output prefaced with a slew of logs relating to metamorphic constants being initialized to non-default values. This can be seen here (note that there is some logged added to `init()` in pkg/util/constants.go):
```
$ bazel test //pkg/kv/kvserver:kvserver_test --test_env=GOTRACEBACK=all --test_filter=TestDecommission --test_arg -test.count=1 --test_arg -test.v --test_sharding_strategy=disabled --test_output all --cache_test_results=no

INFO: Invocation ID: 84f2864c-0830-485e-8bf7-008da04b08c2
INFO: Analyzed target //pkg/kv/kvserver:kvserver_test (0 packages loaded, 0 targets configured).
INFO: Found 1 test target...
INFO: From Testing //pkg/kv/kvserver:kvserver_test:
==================== Test output for //pkg/kv/kvserver:kvserver_test:

-- metamorphic build: true --
initialized metamorphic constant "span-reuse-rate" with value 74
initialized metamorphic constant "parse-json-impl" with value 1
initialized metamorphic constant "disable-checksstconflicts-range-key-masking" with value true
initialized metamorphic constant "addsst-rewrite-concurrency" with value 8
initialized metamorphic constant "addsst-rewrite-forced" with value true
initialized metamorphic constant "raft-log-truncation-clearrange-threshold" with value 294751
initialized metamorphic constant "default-batch-bytes-limit" with value 22062
initialized metamorphic constant "datum-row-converter-batch-size" with value 1
initialized metamorphic constant "row-container-rows-per-chunk-shift" with value 1
initialized metamorphic constant "zig-zag-joiner-batch-size" with value 1
initialized metamorphic constant "merge-joiner-groups-buffer" with value 8
initialized metamorphic constant "direct-scans-enabled" with value true
initialized metamorphic constant "ColIndexJoin-batch-size" with value 2309581
initialized metamorphic constant "ColIndexJoin-using-streamer-batch-size" with value 3324031
initialized metamorphic constant "max-batch-byte-size" with value 5563024
initialized metamorphic constant "parallel-scan-result-threshold" with value 2328
initialized metamorphic constant "copy-batch-size" with value 16882
initialized metamorphic constant "async-IE-result-channel-buffer-size" with value 13
initialized metamorphic constant "use-index-lookup-for-descriptors-in-database" with value false
initialized metamorphic constant "copy-fast-path-enabled-default" with value false
initialized metamorphic constant "multiple_active_portals_enabled" with value true
initialized metamorphic constant "write-metadata-sst" with value true
initialized metamorphic constant "backup.write_metadata_with_external_ssts.enabled" with value false
initialized metamorphic constant "backup-read-manifest" with value true
initialized metamorphic constant "changefeed.fast_gzip.enabled" with value false
initialized metamorphic constant "changefeed.new_webhook_sink_enabled" with value true
initialized metamorphic constant "changefeed.new_pubsub_sink_enabled" with value true

-- metamorphic build: false --
I230420 03:31:50.151970 1 (gostd) rand.go:243  [-] 1  random seed: 3912943573250159482
=== RUN   TestDecommission
    test_log_scope.go:161: test logs captured to: /home/asarkesian/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestDecommission1091475653
    test_log_scope.go:79: use -show-logs to present logs inline
    client_raft_test.go:3389: -- test log scope end --
--- PASS: TestDecommission (19.33s)
PASS
```

As we can see  this output shows both `metamorphic build: true` (with a list of initialized constants) _and_ `metamorphic build: false`, which should not be possible as Go guarantees that a module's `init()` can only be called once. Upon some investigation, it seems that when a Go test binary is run via Bazel, after startup (I.e. initialization and first logging of the metamorphic build state and constants), it forks a child process to actually run the test(s), reinitializing the `metamorphicBuild` flag and any constants to the ones actually used in the test run.

This creates some _extremely_ confusing test output.  Without the logging I added above, you would see a bunch of metamorphic constants that may not actually apply to a given execution, or you may see a number of metamorphic constants with no indication of which ones were initialized in the parent process (and thus no longer applying to the run) and which ones were initialized in the child process, as there would be no line break in-between to indicate this.

I'm not sure there is a way to circumvent this, or to detect in an `init()` (or module-initialized variables) the difference between the processes, as the stack frames/binaries/build flags are all identical.  Hence, the right way to fix this should be to log these metamorphic constants only at startup (I.e. `main()` or `TestMain()`) in code that could be built with the `crdb_test` build tag.

**Comment by AlexTalks** [comment]
With the changes in this PR, the resulting output would look like the following:
```
$ bazel test //pkg/kv/kvserver:kvserver_test --test_env=GOTRACEBACK=all --test_filter=TestDecommission --test_arg -test.count=1 --test_arg -test.v --test_sharding_strategy=disabled --test_output all --cache_test_results=no

INFO: Invocation ID: 84f2864c-0830-485e-8bf7-008da04b08c2
INFO: Analyzed target //pkg/kv/kvserver:kvserver_test (0 packages loaded, 0 targets configured).
INFO: Found 1 test target...
INFO: From Testing //pkg/kv/kvserver:kvserver_test:
==================== Test output for //pkg/kv/kvserver:kvserver_test:
I230420 03:31:50.151970 1 (gostd) rand.go:243  [-] 1  random seed: 3912943573250159482
=== RUN   TestDecommission
    test_log_scope.go:161: test logs captured to: /home/asarkesian/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestDecommission1091475653
    test_log_scope.go:79: use -show-logs to present logs inline
    client_raft_test.go:3276: (pid 29024) this is a non-metamorphic build!
    client_raft_test.go:3389: -- test log scope end --
--- PASS: TestDecommission (19.33s)
PASS
```

**Review by yuzefovich — COMMENTED** [review]
The change LGTM. I think it'd be beneficial to get input from @rickystewart or @healthy-pod whether this behavior of Bazel is expected.

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101895#-:-NTv6iZnEKlW-8iCRdMA:b-xf0wu7)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Comment by rickystewart** [comment]
The behavior is expected and is an artifact of how `rules_go` implements this test behavior. The test binary will [fork itself](https://github.com/cockroachdb/rules_go/blob/58cb94707783f529462d4c6d3c698933a1022942/go/tools/bzltestutil/wrap.go#L69).

The change LGTM and should make test log output much more readable.

**Comment by AlexTalks** [comment]
TFTR @rickystewart.  I'll update the commit / PR message with a summary of the what's written here in the comments.

@rickystewart @yuzefovich any thoughts on if I need review from all those triggered by code owners, or are your reviews good in themselves?

**Comment by yuzefovich** [comment]
> any thoughts on if I need review from all those triggered by code owners, or are your reviews good in themselves?

That's definitely is not necessary - Ricky's and mine approval is sufficient.

How did you insert all those `util.LogMetamorphicVariables()` calls? I wonder whether it's worth writing some automation to do that, or perhaps a linter that ensures that all `main_test.go` files have the call?

**Review by rickystewart — APPROVED** [review]

File: pkg/cmd/generate-logictest/templates.go
Comment (rickystewart) [inline pkg/cmd/generate-logictest/templates.go:289]: Nit: Can you make the indentation here consistent with the context?

**Review by rickystewart — CHANGES_REQUESTED** [review]

**Comment by srosenberg** [comment]
@AlexTalks Thanks for the PR! This triggered a memory while debugging a separate issue pertaining to a wrongfully marked test failure, which forced digging deeper into the guts of bazel. Indeed, each test binary is executed twice which is a rather undesirable side-effect [1].

[1] https://cockroachlabs.slack.com/archives/C023S0V4YEB/p1668025342225139?thread_ts=1667870276.956819&cid=C023S0V4YEB

**Comment by srosenberg** [comment]
> I wonder whether it's worth writing some automation to do that, or perhaps a linter that ensures that all main_test.go files have the call?

Right, automation would be preferred. Some packages don't even have a `TestMain` afaik.

**Comment by AlexTalks** [comment]
@yuzefovich you are absolutely correct that we should ensure this is called in tests with a linter rule, though I'm not sure who can help with that - @rickystewart any thoughts?

As mentioned above, there might be some issues with the tests that don't have a `TestMain(..)`... discussion of how to find those, how the ones already changed were found, and potential solutions is started on [Internal Slack Thread](https://cockroachlabs.slack.com/archives/C019A8NANH0/p1682554340385469)

**Comment by AlexTalks** [comment]
This got split into #102474, #102486, and (??) TBD PRs that actually solve the logging issue by calling the `LogMetamorphicVariables(..)` in the minimal transitive closure set of packages that depend on  `//pkg/util/metamorphic`, using code generation and/or linting.

**Pull request closed without merging**
